File: packages/node/scripts/verify-canonical-schema.js
const { readFileSync } = require('node:fs');
const { resolve } = require('node:path');

const repositorySchema = resolve(__dirname, '../../../schemas/log-event.schema.json');
const packagedSchema = resolve(__dirname, '../src/sdk/log-event.schema.json');

const normalize = (filePath) => JSON.stringify(JSON.parse(readFileSync(filePath, 'utf8')));

if (normalize(repositorySchema) !== normalize(packagedSchema)) {
  throw new Error(
    'Node packaged log-event schema differs from schemas/log-event.schema.json. Copy the canonical schema before building.',
  );
}

console.log('Canonical Node schema verified.');
process.exitCode = 0;

File: packages/python/scripts/verify_canonical_schema.py
"""Fail if the packaged schema drifts from the monorepo canonical schema."""

from __future__ import annotations

import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[3]
PACKAGED = Path(__file__).resolve().parents[1] / "dt3_sdk" / "log-event.schema.json"
CANONICAL = ROOT / "schemas" / "log-event.schema.json"


def _normalize(path: Path) -> str:
    return json.dumps(json.loads(path.read_text(encoding="utf-8")), sort_keys=True)


def main() -> int:
    if not CANONICAL.is_file():
        print(f"Canonical schema missing: {CANONICAL}", file=sys.stderr)
        return 1
    if not PACKAGED.is_file():
        print(f"Packaged schema missing: {PACKAGED}", file=sys.stderr)
        return 1
    if _normalize(CANONICAL) != _normalize(PACKAGED):
        print(
            "Python packaged log-event schema differs from schemas/log-event.schema.json. "
            "Copy the canonical schema into packages/python/dt3_sdk/ before releasing.",
            file=sys.stderr,
        )
        return 1
    print("Canonical Python schema verified.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
